### PR TITLE
[MAINTENANCE] General cleanup/refactor of `DataAssistantResult`

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- General refactors in preparation for `OnboardingDataAssistant` plotting

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 1]
```diff
@@ -38,6 +39,8 @@
 )
 from great_expectations.types import ColorPalettes, Colors, SerializableDictDot
 
+ColumnDataFrame = namedtuple("ColumnDataFrame", ["column", "df"])
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:42]: Just to provide some more clarity, let's associate our column-df tuple with names.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 2]
```diff
@@ -348,7 +351,7 @@ def apply_theme(
         Returns:
             A list of Altair charts with the theme applied
         """
-        theme: Dict[str, Any] = DataAssistantResult._get_theme(theme=theme)
+        theme = DataAssistantResult._get_theme(theme=theme)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:354]: Already declared above so the type annotation is not necessary.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 3]
```diff
@@ -440,15 +443,15 @@ def get_expect_domain_values_to_be_between_chart(
             column
             for column in df.columns
             if column
-            not in [
+            not in {
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:446]: Very minor but set vs list lookup

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 4]
```diff
@@ -1628,16 +1631,64 @@ def _plot_column_domain_charts(
         plot_mode: PlotMode,
         sequential: bool,
     ) -> Tuple[List[alt.VConcatChart], List[alt.Chart]]:
+        column_based_expectation_configurations_by_type: Dict[
+            str, List[ExpectationConfiguration]
+        ] = self._filter_expectation_configurations_by_column_type(
+            expectation_configurations, include_column_names, exclude_column_names
+        )
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1638]: Helper method to filter relevant column-based expectation configurations from the list - the result is a dictionary with keys representing expectation name and values representing the list of relevant configs.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 5]
```diff
@@ -1628,16 +1631,64 @@ def _plot_column_domain_charts(
...
+            for expectation_configuration in column_based_expectation_configurations:
+                return_chart: alt.Chart = (
+                    self._create_return_chart_for_column_domain_expectation(
+                        expectation_configuration=expectation_configuration,
+                        attributed_metrics=attributed_metrics_by_column_domain,
+                        plot_mode=plot_mode,
+                        sequential=sequential,
+                    )
+                )
+                return_charts.append(return_chart)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1669]: For each column-based expectation, create a layer chart (this is only one for the `VolumeDataAssistant` but will be more numerous for the `OnboardingDataAssistant`). I've also moved the `_create_return_chart` calls to be within the same loop to save on work.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 6]
```diff
@@ -1711,11 +1748,10 @@ def _chart_domain_values(
...
         ] = None
         chart: Optional[alt.Chart] = None
-        if plot_mode is PlotMode.PRESCRIPTIVE:
-            if metric_name in implemented_metrics:
+        if metric_name in implemented_metrics:
+            if plot_mode is PlotMode.PRESCRIPTIVE:
                 plot_impl = self.get_expect_domain_values_to_be_between_chart
-        elif plot_mode is PlotMode.DESCRIPTIVE:
-            if metric_name in implemented_metrics:
+            elif plot_mode is PlotMode.DESCRIPTIVE:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1750]: Flip around conditionals to save on some code (don't need to use `if metric_name in implemented_metrics` both times).

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 7]
```diff
@@ -1824,21 +1858,21 @@ def _chart_column_values(
...
                 ],
                 alt.VConcatChart,
             ]
         ] = None
         display_chart: Optional[alt.VConcatChart] = None
-        if plot_mode is PlotMode.PRESCRIPTIVE:
-            if metric_name == "column_distinct_values_count":
+
+        if metric_name == "column_distinct_values_count":
+            if plot_mode is PlotMode.PRESCRIPTIVE:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1864]: I believe this is functionally equivalent but please call me out if not!

```python
        # OLD 
        if plot_mode is PlotMode.PRESCRIPTIVE:
            if metric_name == "column_distinct_values_count":
                plot_impl = (
                    self.get_interactive_detail_expect_column_values_to_be_between_chart
                )
        else:
            if metric_name == "column_distinct_values_count":
                plot_impl = self.get_interactive_detail_multi_chart

       # NEW
        if metric_name == "column_distinct_values_count":
            if plot_mode is PlotMode.PRESCRIPTIVE:
                plot_impl = (
                    self.get_interactive_detail_expect_column_values_to_be_between_chart
                )
            else:
                plot_impl = self.get_interactive_detail_multi_chart
```

Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1864]: Same note with the similar refactor above!

**Review by alexsherstinsky — APPROVED** [review]
LGTM

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py (hunk 1 above)
Comment (Shinnnyshinshin) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:42]: really really like this adjustment

**Review by Shinnnyshinshin — APPROVED** [review]
love it. Thank you @cdkini

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py (hunk 1 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:42]: Love it!

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 8]
```diff
@@ -564,7 +567,7 @@ def get_interactive_detail_multi_chart(
         batch_name: str = "batch"
         batch_identifiers: List[str] = [
             column
-            for column in column_dfs[0][1].columns
+            for column in column_dfs[0].df.columns
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:570]: ❤️

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py [hunk 9]
```diff
@@ -1769,25 +1801,20 @@ def _create_return_chart_for_column_domain_expectation(
...
-        metric_configuration: dict = expectation_configuration.meta["profiler_details"][
-            "metric_configuration"
-        ]
+        profiler_details: dict = expectation_configuration.meta["profiler_details"]
+        metric_configuration: dict = profiler_details["metric_configuration"]
         domain_kwargs: dict = metric_configuration["domain_kwargs"]
+        column_name: str = domain_kwargs["column"]
 
-        domain = domains_by_column_name[domain_kwargs["column"]]
+        domain = domains_by_column_name[column_name]
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1809]: 🤦🏻

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py (hunk 7 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:1864]: This looks good, but as we continue to add more metrics/expectations we will need possibly rethink if the metric names can get pulled from `EXPECTATION_METRIC_MAP` or if this get's pushed down into child classes.

**Review by NathanFarmer — APPROVED** [review]
LGTM!

**Pull request merged**